src/home/homed-home.c

```diff
@@ -1209,6 +1222,35 @@ static int home_start_work(Home *h, const char *verb, UserRecord *hr, UserRecord
                         return r;
         }
 
+        if (blobs) {
+                const char *blob_filename = NULL;
+                void *fd_ptr;
+                size_t i = 0;
+
+                blob_fd_arr = new(int, hashmap_size(blobs));
+                if (!blob_fd_arr)
+                        return -ENOMEM;
+
+                /* homework needs to be able to tell the difference between blobs being null
+                 * (the fdmap field is completely missing) and it being empty (the field is an
+                 * empty object) */
+                r = json_variant_new_object(&fdmap, NULL, 0);
+                if (r < 0)
+                        return r;
+
+                HASHMAP_FOREACH_KEY(fd_ptr, blob_filename, blobs) {
+                        r = json_variant_set_field_integer(&fdmap, blob_filename, PTR_TO_FD(fd_ptr));
+                        if (r < 0)
+                                return r;
+
+                        blob_fd_arr[i++] = PTR_TO_FD(fd_ptr);
+                }
```

Comment:
so i am pretty sure we should just stay close to the LISTEN_FDS= contract. or in other words, pass the fds to the child from fd 3 on, densely packed. e.g. if you want to pass 34 fds to a process you fork off, then you pass them as fds 3…36. And we could even set $LISTEN_FDS env var one day, if we like. And then, inside the json we'd just pass the index into the densely packed array around, rather than the fd number itself.

(I mean, this will always be an array that just contains 0, 1, 2, 3, … strictly ordered, but that's fine, i think it makes sense to still do this, so that later on we can possibly pass more fds over, at which point it starts making sense to maintain this as indexes).

hence, build the array as you already do, and then maybe reuse shift_fds() from core/exec-invoke.c to pack them densely into fd 3,4,5,… right before we invoke homework as child process.

Reply:
I think I've done it. Please take a look